app/v1/data.py

```diff
@@ -17,9 +17,82 @@
 log = structlog.get_logger()
 
 
+DATA_TYPES = Literal["csv", "feather", "feather_direct", "json"]
+
 router = APIRouter()
 
 
+def _read_sql_bytes(con, sql: str, parameters) -> io.BytesIO:
+    """Execute SQL and return BytesIO object with byte data."""
+    sink = io.BytesIO()
+
+    batch_iterator = con.execute(sql, parameters=parameters).fetch_record_batch(
+        chunk_size=1000
+    )
+    with pa.ipc.new_file(sink, batch_iterator.schema) as writer:
+        for rb in batch_iterator:
+            writer.write_batch(rb)
+
+    sink.seek(0)
+
+    return sink
+
+
+def _sql_to_response(
+    con, sql: str, type: DATA_TYPES, parameters: list[Any] = []
+) -> Any:
+    # read data in feather format and return it directly in response
+    # NOTE: should be the fastest in theory, but is really slow for unknown reasons
+    if type == "feather_direct":
+        bytes_io = _read_sql_bytes(con, sql, parameters=parameters)
+        # NOTE: this approach is much faster than `StreamingResponse(bytes_io, ...`, how is it possible?
+        # response = StreamingResponse(
+        #     iter([bytes_io.getvalue()]), media_type="application/octet-stream"
+        # )
+        response = StreamingResponse(bytes_io, media_type="application/octet-stream")
+        response.headers["Content-Disposition"] = "attachment; filename=test.feather"
+        return response
+
+    # read data into dataframe and then convert to feather
+    elif type == "feather":
+        bytes_io = io.BytesIO()
+        df = con.execute(sql, parameters=parameters).fetch_df()
+        write_feather(df, bytes_io)
+
+        response = StreamingResponse(
+            iter([bytes_io.getvalue()]), media_type="application/octet-stream"
+        )
+        response.headers["Content-Disposition"] = "attachment; filename=test.feather"
```

Comment:
should we have a subroutine for this `StreamingResponse` stuff? looks about the same here and above

Reply:
also another `test.feather` here

Reply:
DRYed, thanks